apps/enterprise-knowledge-retrieval/enterprise_knowledge_retrieval.ipynb

```diff
@@ -0,0 +1,1750 @@
+    "    input_variables=[\"input\", \"intermediate_steps\", \"history\"]\n",
+    ")\n",
+    "\n",
+    "# Initiate the memory with k=2 to keep the last two turns\n",
+    "# Provide the memory to the agent\n",
+    "memory = ConversationBufferWindowMemory(k=2)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 122,
+   "id": "34de07d2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "output_parser = CustomOutputParser()\n",
+    "\n",
+    "llm = ChatOpenAI(temperature=0,openai_organization='org-l89177bnhkme4a44292n5r3j')\n",
+    "\n",
+    "# LLM chain consisting of the LLM and a prompt\n",
+    "llm_chain = LLMChain(llm=llm, prompt=prompt)\n",
+    "\n",
+    "tool_names = [tool.name for tool in tools]\n",
+    "agent = LLMSingleActionAgent(\n",
+    "    llm_chain=llm_chain, \n",
+    "    output_parser=output_parser,\n",
+    "    stop=[\"\\nObservation:\"], \n",
+    "    allowed_tools=tool_names\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 123,
+   "id": "d3603d58",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "agent_executor = AgentExecutor.from_agent_and_tools(agent=agent, tools=tools, verbose=True, memory=memory)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 124,
+   "id": "6bfb594b",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n",
+      "\n",
+      "\u001b[1m> Entering new AgentExecutor chain...\u001b[0m\n",
+      "\u001b[32;1m\u001b[1;3mThought: I should use the Search tool to find information about Thomas Dolby\n",
+      "Action: Search\n",
+      "Action Input: \"What is Thomas Dolby known for?\"\u001b[0m\n",
+      "\n",
+      "Observation:\u001b[36;1m\u001b[1;3mThomas Dolby is known for his 1982 hit \"She Blinded Me With Science\" and his career as a British musician and computer designer. He has also released several studio albums and EPs. Source: Thomas Dolby Wikipedia page.\u001b[0m\u001b[32;1m\u001b[1;3mI now know what Thomas Dolby is known for.\n",
+      "Final Answer: Thomas Dolby is known for his hit song \"She Blinded Me With Science\" and his career as a British musician and computer designer.\u001b[0m\n",
+      "\n",
+      "\u001b[1m> Finished chain.\u001b[0m\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "'Thomas Dolby is known for his hit song \"She Blinded Me With Science\" and his career as a British musician and computer designer.'"
+      ]
+     },
+     "execution_count": 124,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "agent_executor.run(f1_query)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 125,
+   "id": "ba65b7e3",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
```

Comment:
I got errors on all my of first attempts here. Looks like it breaks if doesn't try searching. Only when I started asking it Wikipedia-style questions did it start working.

Could be worth:
- Adding a note to explain this
- Adding logic to handle non-search outcomes

Reply:
I've tuned the prompt now, I should have done this from the start - if the tool doesn't answer the question it needs an 
"ELSE" otherwise it loops forever. Now adjusted.